### PR TITLE
add etcd lease design doc

### PR DESCRIPTION
**Pull request by StupidHod**
@AdoHe  @xiang90 @hongchaodeng

### CODE REVIEW

**Comment by xiang90** [comment]
let's separate them into two prs. design -> interface/impl.

**Comment by StupidHod** [comment]
@xiang90 Ok, tks for your advice.

File: Documentation/EtcdLease.md [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+## keepAlive function
+1. It will create a Lease Object, and keep the lease alive.
+2. It will add the created lease object to the keepAlives map.
+3. The background scheduler will send keep alive requests for the added lease when it approaches to the nextKeepAliveTime.
+
+## keepAliveExecutor function
+1. This function is called periodically by `keepAliveSchedule`
+2. The lease may expire as:
+-   The etcd respond overtime.
```
Comment (xiang90) [inline Documentation/EtcdLease.md]: this format is not quite right. click view to see the rendered view.

File: Documentation/EtcdLease.md [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+3. call the on onLeaseExpired method for these leases.
+
+## keepAliveResponseStreamObserver interface instance
+1. The StreamObserver is the stream etcd uses to send responses.
+2. The gRPC client calls onNext when etcd server delivers a response.
+3. The onNext function will call `processKeepAliveRespond`
+
+## processKeepAliveRespond function
+1. It will get the lease instance based on the leaseID.
+2. if repond ttl < 0, it remove the lease from keepAlives map and call `onLeaseExpired`
```
Comment (xiang90) [inline Documentation/EtcdLease.md]: if TTL < 0 (remove repond)

File: Documentation/EtcdLease.md [hunk 3]
```diff
@@ -0,0 +1,41 @@
+# Overview
+The EtcdLease implement interface `grant` and `revoke` to new a lease and revoke a lease to etcd.
```
Comment (xiang90) [inline Documentation/EtcdLease.md]: please add a empty line between title and the content. (also for the following sections)

**Comment by xiang90** [comment]
LGTM after addressing the comments. @AdoHe @skyao  Please feel free to merge this in (you are both in China time zone :) )

Thanks a lot.

File: Documentation/EtcdLease.md [hunk 4]
```diff
@@ -2,40 +2,57 @@
 The EtcdLease implement interface `grant` and `revoke` to new a lease and revoke a lease to etcd.
 It can manage the leases by send keep alive request for leases.
 
+##grant function
```
Comment (xiang90) [inline Documentation/EtcdLease.md]: space between `##` and grant

File: Documentation/EtcdLease.md [hunk 5]
```diff
@@ -2,40 +2,57 @@
...
+1. The function build leaseGrantRequest with ttl value.
+2. It call respond gRPC interface with leaseGrantRequest to build a new lease in etcd.
+
+##revoke function
```
Comment (xiang90) [inline Documentation/EtcdLease.md]: same here

**Comment by xiang90** [comment]
@StupidHod Please squash all commits into one. Thanks. I will get it merged.

File: Documentation/EtcdLease.md [hunk 6]
```diff
@@ -0,0 +1,69 @@
+# Overview
+
+The EtcdLease implement interface `grant` and `revoke` to new a lease and revoke a lease to etcd.
```
Comment (hongchaodeng) [inline Documentation/EtcdLease.md]: The EtcdLease interface provides methods to grant, revoke, and keepalive leases.

File: Documentation/EtcdLease.md [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+It can manage the leases by send keep alive request for leases.
```
Comment (hongchaodeng) [inline Documentation/EtcdLease.md]: Remove this line.

**Comment by hongchaodeng** [comment]
LGTM after squashing commits. Thanks!

**Comment by AdoHe** [comment]
@StupidHod please squash all commits into one.

**Comment by StupidHod** [comment]
@AdoHe Ok, I have squashed all commits into one.

**Pull request merged**